packages/decoder/lib/utils.ts

```diff
@@ -0,0 +1,181 @@
+      contractNode.nodeType === "ContractDefinition" &&
+      (contractNode.name === contract.contractName ||
+        contractNode.name === contract.contract_name)
+  );
+}
+
+export function makeContext(
+  contract: ContractObject,
+  node: Codec.Ast.AstNode | undefined,
+  isConstructor = false
+): Codec.Contexts.DecoderContext {
+  const abi = Codec.Abi.Utils.schemaAbiToAbi(contract.abi);
+  const binary = isConstructor ? contract.bytecode : contract.deployedBytecode;
+  const hash = Codec.Conversion.toHexString(
+    Codec.Evm.Utils.keccak256({
+      type: "string",
+      value: binary
+    })
+  );
+  return {
+    context: hash,
+    contractName: contract.contractName,
+    binary,
+    contractId: node ? node.id : undefined,
+    contractKind: contractKind(contract, node),
+    isConstructor,
+    abi: Codec.Abi.Utils.computeSelectors(abi),
+    payable: Codec.Abi.Utils.abiHasPayableFallback(abi),
+    hasFallback: Codec.Abi.Utils.abiHasFallback(abi),
+    compiler: contract.compiler
+  };
+}
+
+//attempts to determine if the given contract is a library or not
+function contractKind(
+  contract: ContractObject,
+  node?: Codec.Ast.AstNode
+): Codec.ContractKind {
+  //first: if we have a node, use its listed contract kind
+  if (node) {
+    return node.contractKind;
+  }
+  //next: check the contract kind field on the contract object itself, if it exists.
+  //however this isn't implemented yet so we'll skip it.
+  //next: if we have no direct info on the contract kind, but we do
+  //have the deployed bytecode, we'll use a HACK:
+  //we'll assume it's an ordinary contract, UNLESS its deployed bytecode begins with
+  //PUSH20 followed by 20 0s, in which case we'll assume it's a library
+  //(note: this will fail to detect libraries from before Solidity 0.4.20)
+  if (contract.deployedBytecode) {
+    const pushAddressInstruction = (
+      0x60 +
+      Codec.Evm.Utils.ADDRESS_SIZE -
+      1
+    ).toString(16); //"73"
+    const libraryString =
+      "0x" + pushAddressInstruction + "00".repeat(Codec.Evm.Utils.ADDRESS_SIZE);
+    return contract.deployedBytecode.startsWith(libraryString)
+      ? "library"
+      : "contract";
+  }
+  //finally, in the absence of anything to go on, we'll assume it's an ordinary contract
+  return "contract";
+}
+
+//Function for wrapping a value as an ElementaryValue
+//WARNING: this function does not check its inputs! Please check before using!
+//How to use:
+//numbers may be BN, number, or numeric string
+//strings should be given as strings. duh.
+//bytes should be given as hex strings beginning with "0x"
+//addresses are like bytes; checksum case is not required
+//booleans may be given either as booleans, or as string "true" or "false"
+//[NOTE: in the future this function will:
+//1. check its inputs,
+//2. take a slightly different input format,
+//3. also be named differently and... it'll be different :P ]
+export function wrapElementaryViaDefinition(
```

Comment:
Hm, so I was originally going to comment that I don't think this function should be in `decoder`, I think it should be `codec`.  Remember that this is eventually going to be expanded into a much broader set of wrapper functions to be used for encoding.

OTOH, arguably that belongs in the high-level interface, so `decoder` is the right place after all.  Especially if wrapping is going to eventually include ENS integration!

(This does bring up once again that circular dependency problem I've been mentioning, with `contract` and `decoder` both depending on each other, since currently ENS resolution lives in `contract`...)

Reply:
(Maybe I should ping @eggplantzzz here?)

Reply:
Mostly what I did, when deciding to move things to the decoder, is if they are used only in the decoder, I made sure they were also defined in the decoder.